controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,448 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mirroring
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	ocsv1alpha1 "github.com/red-hat-storage/ocs-operator/api/v4/v1alpha1"
+	providerClient "github.com/red-hat-storage/ocs-operator/services/provider/api/v4/client"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/storageclusterpeer"
+	controllers "github.com/red-hat-storage/ocs-operator/v4/controllers/storageconsumer"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
+
+	"github.com/go-logr/logr"
+	rookCephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	// internalKey is a special key for client-mapping-config to establish mirroring between blockPools for internal mode
+	internalKey        = "internal"
+	mirroringFinalizer = "mirroring.ocs.openshift.io"
+)
+
+// MirroringReconciler reconciles a Mirroring fields for Ceph Object(s)
+// nolint:revive
+type MirroringReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	log                 logr.Logger
+	ctx                 context.Context
+	clientMappingConfig *corev1.ConfigMap
+	storageClusterPeer  *ocsv1.StorageClusterPeer
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MirroringReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	if err := mgr.GetCache().IndexField(
+		context.TODO(),
+		&ocsv1alpha1.StorageConsumer{},
+		"annotations",
+		util.AnnotationIndexFieldFunc,
+	); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer on StorageConsumer for annotations: %v", err)
+	}
+
+	if err := mgr.GetCache().IndexField(
+		context.TODO(),
+		&ocsv1alpha1.StorageConsumer{},
+		"clientID",
+		util.StorageConsumerClientIDIndexFieldFunc,
+	); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer on StorageConsumer for clientID: %v", err)
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ConfigMap{}, builder.WithPredicates(util.NamePredicate(storageclusterpeer.StorageClientMappingConfigName))).
+		Watches(&ocsv1.StorageClusterPeer{}, &handler.EnqueueRequestForObject{}).
+		Watches(&ocsv1alpha1.StorageConsumer{}, &handler.EnqueueRequestForObject{}).
```

Comment:
By design, this controller is supposed to handle configmap requests. we should not allow other object kind to be enqueued